packages/devtools_app/lib/src/screens/inspector_v2/inspector_tree_controller.dart

```diff
@@ -187,78 +183,99 @@ class InspectorTreeController extends DisposableController
-    _searchTarget = searchTarget;
-    refreshSearchMatches();
+  /// All the rows that should be displayed in the tree.
+  ///
+  /// The rows can be updated with a call to [_updateRows].
+  ValueListenable<List<InspectorTreeRow?>> get rowsInTree => _rowsInTree;
+  final _rowsInTree = ListValueNotifier<InspectorTreeRow?>([]);
+
+  /// Map from node to the index for that node's row in [rowsInTree].
+  final _nodeToRowIndex = <InspectorTreeNode, int>{};
+
+  /// Rebuilds the tree and updates [rowsInTree] with the new values.
+  ///
+  /// If [updateSearchableRows] is true, also updates [_searchableCachedRows]
+  /// with the new values.
+  ///
+  /// TODO(elliette): Consider only updating an [InspectorTreeNode]'s branch
+  /// when it is marked as dirty, instead of the entire tree. See:
+  /// https://github.com/flutter/devtools/issues/7980
+  void _updateRows({
+    InspectorTreeNode? node,
+    bool updateSearchableRows = false,
+  }) {
+    node ??= root;
+    if (node == null) return;
+
+    final rows = _buildRows(node);
+    _rowsInTree.replaceAll(rows);
+    _rowsInTree.notifyListeners();
+
+    // Build the reverse node-to-index map for faster lookups:
+    for (int i = 0; i < _rowsInTree.value.length; i++) {
+      final row = _rowsInTree.value[i];
+      final node = row?.node;
+      if (node != null) {
+        _nodeToRowIndex[node] = i;
+      }
+    }
+
+    if (updateSearchableRows) {
+      _searchableCachedRows.clear();
+      _searchableCachedRows.addAll(rows);
+    }
   }
 
-  // TODO: we should add a listener instead that clears the cache when the
-  // root is marked as dirty.
-  void _maybeClearCache() {
-    final rootLocal = root;
-    if (rootLocal != null && rootLocal.isDirty) {
-      cachedRows.clear();
+  /// Resets the state if the root has been marked as dirty.
+  void _handleDirtyNode(InspectorTreeNode node) {
+    if (node == root) {
       _cachedSelectedRow = null;
-      rootLocal.isDirty = false;
       lastContentWidth = null;
+      _updateRows();
     }
   }
 
-  void _populateSearchableCachedRows() {
-    _searchableCachedRows.clear();
-    for (int i = 0; i < numRows; i++) {
-      _searchableCachedRows.add(getCachedRow(i));
-    }
+  void setSearchTarget(SearchTargetType searchTarget) {
+    _searchTarget = searchTarget;
+    refreshSearchMatches();
   }
 
-  InspectorTreeRow? getCachedRow(int index) {
+  InspectorTreeRow? getRowAtIndex(int index) {
     if (index < 0) return null;
 
-    _maybeClearCache();
-    while (cachedRows.length <= index) {
-      cachedRows.add(null);
+    if (_rowsInTree.value.length > index) {
+      return _rowsInTree.value[index];
     }
-    cachedRows[index] ??= root?.getRow(index);
-
-    final cachedRow = cachedRows[index];
-    cachedRow?.isSearchMatch =
-        _searchableCachedRows.safeGet(index)?.isSearchMatch ?? false;
-
-    if (cachedRow?.isSelected == true) {
-      _cachedSelectedRow = cachedRow;
-    }
-    return cachedRow;
+    return null;
```

Comment:
this can be condensed:
```
return _rowsInTree.value.safeGet(index)
```

Also consider renaming method to avoid starting method names with get. Perhaps `rowAtIndex`?

Reply:
Ah nice helper function! Changed all the various `getRow....` methods to just `row...`